### PR TITLE
Fix PartsSplitter

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Bug Fix

### Changelog entry:
When some index columns are not loaded into memory for some parts of a *MergeTree table, queries with `FINAL` might produce wrong results. Now we explicitly choose only the common prefix of index columns for all parts to avoid this issue.

---

Related: #62173 #62192 #62252 

cc @nikitamikhaylov @jkartseva

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 4e11cd5be9f6a3b949028bfaecd0f91d65af025a with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62268/4e11cd5be9f6a3b949028bfaecd0f91d65af025a/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by nickitat** [comment]
apparently problem is coming from recent optimisation of not loading some index columns. splitting on layers was incorrect, for one part instead of range (0, 4) we took only (0, 2). I noticed that for it we use only first column value as the index value, while the index consists of 3 columns. then I set `primary_key_ratio_of_unique_prefix_values_to_skip_suffix_columns=1` and it fixed the test.

File: tests/integration/test_final_bug_with_pk_columns_loading/test.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import pytest
```
Comment (alexey-milovidov) [inline tests/integration/test_final_bug_with_pk_columns_loading/test.py:1]: This could be a functional test.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

File: tests/integration/test_final_bug_with_pk_columns_loading/test.py (hunk 1 above)
Comment (nikitamikhaylov) [reply tests/integration/test_final_bug_with_pk_columns_loading/test.py:1]: But the test restarts ClickHouse server.

Comment (nickitat) [reply tests/integration/test_final_bug_with_pk_columns_loading/test.py:1]: yep, I'm not aware of a way to trigger index optimisation w/o server restart

**Comment by nickitat** [comment]
as some of us already noticed, `00147_global_in_aggregate_function` failed in CI with this fix.

https://s3.amazonaws.com/clickhouse-test-reports/62268/378d330d9dfa289c413f80c2addaf6dee5503093/stateful_tests__tsan_.html

though it seems there is a different issue, because this test fails almost in 100% cases with the following settings:

``` sql
set merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability=0;
set max_block_size=69749, max_threads=32;
```

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**